### PR TITLE
OADP-599: Auto create s3 restic secrets for volsync from BSLs

### PR DESCRIPTION
**Pull request by savitharaghunathan**

### CODE REVIEW

File: controllers/datamover.go
Comment (eemcmullan) [inline controllers/datamover.go:273]: should this be a const value?

File: controllers/datamover.go
Comment (eemcmullan) [inline controllers/datamover.go:289]: Since this only checks for a password, should the error reflect that? Something like "no password supplied in restic secret." wdyt?

File: controllers/datamover.go
Comment (eemcmullan) [inline controllers/datamover.go:272]: NIT: then

File: controllers/datamover.go
Comment (eemcmullan) [inline controllers/datamover.go:193]: should this be the other way around? returning true if there is a supplied password

File: api/v1alpha1/oadp_types.go
Comment (eemcmullan) [inline api/v1alpha1/oadp_types.go:168]: NIT: supplied*

File: controllers/datamover.go
Comment (savitharaghunathan) [reply controllers/datamover.go:193]: ha ha 100% :D what was I thinking. Thanks Emily

File: controllers/datamover.go
Comment (savitharaghunathan) [reply controllers/datamover.go:269]: no. I dont see any use case where it needs to be deleted or updated. These stay as long as we have oadp operator running.

Comment (savitharaghunathan) [reply controllers/datamover.go:269]: Do you have any use case in mind?

File: controllers/datamover.go
Comment (savitharaghunathan) [reply controllers/datamover.go]: changed condition

**Comment by savitharaghunathan** [comment]
/hold
todo: Add plugin support for passing restic secret name in VSB & VSR CR
Add support for restic secret names in snap mover

File: controllers/datamover.go
Comment (shubham-pampattiwar) [inline controllers/datamover.go:222]: I think this is not always the case for s3 repo path. The URL format of a bucket is either of two options: http://s3.amazonaws.com/[bucket_name]/ http://[bucket_name].s3.amazonaws.com/, not entirely sure about this.

File: controllers/datamover.go
Comment (shubham-pampattiwar) [inline controllers/datamover.go:226]: `resticsecret` <-> `restic-secret`

File: controllers/datamover.go
Comment (shubham-pampattiwar) [inline controllers/datamover.go:229]: Please add the labels as a string constant in `oadp_types.go`

File: controllers/datamover.go
Comment (savitharaghunathan) [reply controllers/datamover.go:222]: @shubham-pampattiwar I am using the format that we use in restic repo secret. Which format would you prefer?

File: controllers/datamover.go
Comment (eemcmullan) [reply controllers/datamover.go:273]: Gotcha. Just wanted to double check

File: controllers/datamover.go
Comment (savitharaghunathan) [reply controllers/datamover.go:236]: looks like I am checking that already in L#191. I will remove this redundant check

File: controllers/datamover.go
Comment (kaovilai) [reply controllers/datamover.go:222]: https://restic.readthedocs.io/en/stable/030_preparing_a_new_repo.html?highlight=minio#minio-server

Comment (shubham-pampattiwar) [reply controllers/datamover.go:222]: The repo URL is a bit dynamic here to be created by us, IMO lets keep this on the user to supply via a secret.

Comment (savitharaghunathan) [reply controllers/datamover.go]: Made changes to obtain repo from the user

**Review by eemcmullan — APPROVED** [review]

File: config/crd/bases/oadp.openshift.io_dataprotectionapplications.yaml
Comment (dymurray) [inline config/crd/bases/oadp.openshift.io_dataprotectionapplications.yaml:448]: I'm wondering should we move these configs to be:
```
features:
  datamover:
    enable: true
    credentials: <secret>
```

This way if we have more configs we can stick it under a top level `datamover` key?

**Review by shubham-pampattiwar — CHANGES_REQUESTED** [review]
Added a few suggestions.

**Comment by kaovilai** [comment]
Don't forget `make bundle` I expect to see changes in `config/` replicated in `bundle/`

File: controllers/datamover.go
Comment (eemcmullan) [inline controllers/datamover.go:28]: this will be changed to 'dm-credential' right?

Comment (savitharaghunathan) [reply controllers/datamover.go:28]: yes. Submitting a commit right away

Comment (savitharaghunathan) [reply controllers/datamover.go]: done

**Comment by kaovilai** [comment]
@savitharaghunathan please run `make bundle` to duplicate changes to bundle/

And revert changes related to allowing null in CRD

**Comment by savitharaghunathan** [comment]
@kaovilai I did run make bundle. I had to manually put the nullable true back.

**Review by kaovilai — APPROVED** [review]

**Comment by savitharaghunathan** [comment]
@kaovilai Is there any file changes you are expecting to see other than the two that are in the commits?

**Comment by kaovilai** [comment]
Nope. Just misread the changes

**Comment by kaovilai** [comment]
Last
/retest
was 23 hours ago
Resolution to label length related failures were last resolved 15 hours ago
https://github.com/openshift/release/tree/master/ci-operator/config/openshift/oadp-operator

**Comment by kaovilai** [comment]
@hhpatel14 any reason https://github.com/openshift/release/blob/master/ci-operator/config/openshift/oadp-operator/openshift-oadp-operator-master__4.10.yaml is different to 
https://github.com/openshift/release/blob/master/ci-operator/config/openshift/oadp-operator/openshift-oadp-operator-master__4.9.yaml ?

Specifically
```diff
+  openshift_release_golang-1.16:
+    name: builder
+    namespace: ocp
+    tag: rhel-8-golang-1.16-openshift-4.9
-  openshift_release_golang-1.16:
-    name: release
-    namespace: openshift
-    tag: golang-1.16
```

File: controllers/datamover.go
Comment (shubham-pampattiwar) [inline controllers/datamover.go:148]: Maybe rename the function to `CreateDataMoverSecret` ? And Also, not add this here, instead add this in the main OADP Oeprator Batch Reconcile function ?

File: controllers/datamover.go
Comment (shubham-pampattiwar) [inline controllers/datamover.go:226]: Once you move this function to the main batch reconcile, I believe you will need to add the check whether dm is enabled else skip this function's execution altogether.

File: controllers/datamover.go
Comment (shubham-pampattiwar) [inline controllers/datamover.go:259]: As the secret is create by OADP Operator, maybe add the operatort label as well, so that we do not miss on the reconciles via events on these secrets

File: controllers/datamover.go
Comment (shubham-pampattiwar) [inline controllers/datamover.go:303]: Once we move this to Reconcile batch, you will also have to fetch fresh DPA instance.

**Review by shubham-pampattiwar — COMMENTED** [review]
Changes look sane to me, Added some suggestions though. Thank you @savitharaghunathan

File: controllers/datamover.go
Comment (savitharaghunathan) [reply controllers/datamover.go:76]: if `dpa.Spec.Feature` is nil, wouldnt `dpa.Spec.Feature.Datamover` be nil as well?

**Comment by savitharaghunathan** [comment]
@shubham-pampattiwar Thanks for the elaborate review. I updated the func according to the last feedback. When you get a moment can you take a look at it please?

File: controllers/datamover.go
Comment (shubham-pampattiwar) [reply controllers/datamover.go:76]: So currently this might work, but lets say we add more features and Datamover is not the only tech preview feature, then it might right?

Comment (savitharaghunathan) [reply controllers/datamover.go:76]: Makes sense. will add that to the conditional.

Comment (shubham-pampattiwar) [reply controllers/datamover.go:76]: Thank you !

Comment (savitharaghunathan) [reply controllers/datamover.go]: updated

File: controllers/datamover.go
Comment (shubham-pampattiwar) [inline controllers/datamover.go:315]: I think you need `dpa.Spec.Features != nil` as well, else you might hit a nil ptr @savitharaghunathan

Comment (shubham-pampattiwar) [inline controllers/datamover.go:315]: Suggestion: Create a function which has all these checks on dpa and returns to you whether datamover is enabled or not, might make the code more readable and we can also escape from bugs if a dev forgets to adda nil check. WDTY ? @savitharaghunathan

Comment (savitharaghunathan) [reply controllers/datamover.go:315]: on it

File: controllers/datamover.go
Comment (shubham-pampattiwar) [inline controllers/datamover.go:367]: @savitharaghunathan Are we assuming that the given datamover secret will work for the BSL configured ? According to this current code, if I am not wrong, we are creating a dm secret per BSL and password provided in this secret will work for all the BSLs configured ?
For e.g: 
1. OADP has 2 BSLs - AWS and GCP. User supplied one dm secret,this PR will be creating restic secret using the dm secret for both the BSLs, right ? I think we should only create the restic secret for the dm one.
2. OADP has 2 BSLs - Both AWS. Same issue.

**Suggested Solution:**
We add a BSLName and Provider field to Deature -> DataMover/ And we create the restic secret only for the BSL specified here. WDYT ?

Comment (shubham-pampattiwar) [reply controllers/datamover.go:367]: Please correct me if I am wrong here.

File: bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml
Comment (kaovilai) [inline bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml:2]: I just tried running make bundle. These two lines don't exist here. Yes they exists in config/crd/bases/oadp.openshift.io_dataprotectionapplications.yaml but it doesn't belong in bundle/manifests.

Just let `make bundle` do its thing. Hopefully that fixes it?
```suggestion
```

**Review by kaovilai — CHANGES_REQUESTED** [review]

File: bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml
Comment (savitharaghunathan) [reply bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml:2]: Let me try that. Thanks @kaovilai

Comment (savitharaghunathan) [reply bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml]: updated

**Review by shubham-pampattiwar — APPROVED** [review]
LGTM ! Thank you so much @savitharaghunathan !

File: controllers/datamover.go
Comment (dymurray) [inline controllers/datamover.go:212]: We want to update this to check if the key exists at all, right now I get no error if the key isn't set

File: config/crd/bases/datamover.oadp.openshift.io_volumesnapshotbackups.yaml
Comment (dymurray) [inline config/crd/bases/datamover.oadp.openshift.io_volumesnapshotbackups.yaml:48]: I would remove this from the description

**Pull request merged**
